### PR TITLE
Improve ES status health reporting

### PR DESCRIPTION
**Pull request by thbkrkr**
2 first commits are more an optimization:
- Split HTTP and transport certs reconciliation in order to be able to retrieve the cluster health despite an issue with the transport certs.
- Start the observer and start it as soon as possible.

Those after could be enough:
- Initialize health state to unknown in the NewState constructor

Relates to #5330.

### Testing

- Create an ES cluster
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: eight
spec:
  version: 8.0.0
  # http:
  # transport:
  #   tls:
  #     certificate:
  #       secretName: ca-that-does-not-exist
  nodeSets:
  - name: master
    count: 1
    config:
      node.store.allow_mmap: false
```
- Make cluster health yellow
```
# create an index with 1 replica
-XPUT /my-index-1 -d '{"settings":{"number_of_shards":1,"number_of_replicas":1}}'
```
- Break the transport service by configuring a CA Secret that doesn't exist in the cluster

Uncomment `transport.tls.certificate.secretName` and reapply.

- Make cluster health green Again
```
# delete the index
-XDELETE /my-index-1
```

- Cluster should be reported as `green` and not stuck in the `yellow` state, with phase `ApplyingChanges`.

- Do the same by breaking the `http` service. This time the operator will report an `unknown` state as the observer depends on this service, still in phase `ApplyingChanges`.

### CODE REVIEW

**Comment by mtparet** [comment]
Quick question, does it means the operator will also set the status to `unknown` when we excluding an elasticsearch from being managed by the operator using https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-upgrading-eck.html#k8s-beta-to-ga-rolling-restart ?

Because today, if we exclude temporary a ressources, the status is reported is kept also it does not reflect the reality.

**Comment by thbkrkr** [comment]
> does it means the operator will also set the status to unknown when we excluding an elasticsearch from being managed by the operator 

No, it is independent.

> Because today, if we exclude temporary a ressources, the status is reported is kept also it does not reflect the reality.

Could you elaborate on that (maybe in #5330 or a dedicated issue)? What part of the status isn't updated?

**Comment by mtparet** [comment]
> Could you elaborate on that (maybe in https://github.com/elastic/cloud-on-k8s/issues/5330 or a dedicated issue)? What part of the status isn't updated?

I think it's a different issue. 
If we set the annotation `eck.k8s.elastic.co/managed=false` to a resource, the operator will keep the status as previously (say "green" for example) although it does not update anymore the status whatever happens to the elasticsearch cluster.

**Comment by thbkrkr** [comment]
I realized that we can reset the `State` with an empty `ResourcesState`, so before the `ResourcesState` is created and before we have reconciled the services on which it depends on. c370d1e

File: pkg/controller/elasticsearch/driver/driver.go [hunk 1]
```diff
@@ -115,6 +115,10 @@ var _ commondriver.Interface = &defaultDriver{}
 func (d *defaultDriver) Reconcile(ctx context.Context) *reconciler.Results {
 	results := reconciler.NewResult(ctx)
 
+	// reset the Elasticsearch health to 'unknown' so that if reconciliation fails before the observer has had a chance to get it,
+	// we stop reporting a state that may be out of date
+	d.ReconcileState.UpdateElasticsearchPendingUnknownHealth()
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/driver.go:120]: Would it not be easier to change the `NewState` constructor to initialise the state with `Unknown`?

**Review by pebrc — COMMENTED** [review]
I am not sure the split in certificate reconciliation is worth the effort (but I am also not opposed to it). 

Regarding the reset to `unknown` I think we could instead just initialise the reconciliation state with `unknown` health in the reconcile state constructor function. 

Another possible improvement could be to update the status sub resource even if the cluster is unmanaged but that should probably be discussed in an issue first.

File: pkg/controller/elasticsearch/driver/driver.go (hunk 1 above)
Comment (thbkrkr) [reply pkg/controller/elasticsearch/driver/driver.go]: Much better indeed! And this way, we keep the other information (number of nodes, version, ...) that are retrieved from the existing custom resource, therefore still valid.

**Comment by thbkrkr** [comment]
> I am not sure the split in certificate reconciliation is worth the effort (but I am also not opposed to it).

I discussed with barkbay and he is rather in favor of it but will share his opinion after verifying that it goes well with #5328.

File: pkg/controller/common/certificates/reconcile.go [hunk 2]
```diff
@@ -53,9 +51,6 @@ type Reconciler struct {
 // - a Secret containing the public-facing HTTP certificates (same as the internal one, but without the key)
 // If TLS is disabled, self-signed certificates are still reconciled, for simplicity/consistency, but not used.
 func (r Reconciler) ReconcileCAAndHTTPCerts(ctx context.Context) (*CertificatesSecret, *reconciler.Results) {
-	span, _ := apm.StartSpan(ctx, "reconcile_certs", tracing.SpanTypeApp)
```
[the remark this answers is not in the document]
Comment (thbkrkr) [reply pkg/controller/common/certificates/reconcile.go:56]: It is mistake, I'll revert. With the certificate reconciliation split, I start a span in `ReconcileHTTP` so I thought we could remove that one but I forgot all the other uses of `ReconcileCAAndHTTPCerts`.

**Pull request merged**